pico_w/mqtt/mqtt_client/mqtt_client.c

```diff
@@ -0,0 +1,161 @@
+    char temp_str[16];
+    snprintf(temp_str, sizeof(temp_str), "%.2f", temperature);
+    mqtt_publish(mqtt->mqtt_client_inst, "/temperature", temp_str, strlen(temp_str), 0, 0, NULL, NULL);
+}
+static void mqtt_incoming_data_cb(void *arg, const u8_t *data, u16_t len, u8_t flags) {
+    MQTT_CLIENT_DATA_T* mqtt_client = (MQTT_CLIENT_DATA_T*)arg;
+    strncpy(mqtt_client->data, data, len);
+    mqtt_client->len = len;
+    mqtt_client->data[len] = '\0';
+
+    printf("Topic: %s, Message: %s\n", mqtt_client->topic, mqtt_client->data);
+
+    if (strcmp(mqtt->topic, "/led") == 0)
+    {
+        if (strcmp((const char *)mqtt_client->data, "On") == 0)
+            control_led(true);
+        else if  (strcmp((const char *)mqtt_client->data, "Off") == 0)
+            control_led(false);
+    }
+}
+
+static void mqtt_incoming_publish_cb(void *arg, const char *topic, u32_t tot_len) {
+    MQTT_CLIENT_DATA_T* mqtt_client = (MQTT_CLIENT_DATA_T*)arg;
+    strcpy(mqtt_client->topic, topic);
+}
+
+static void mqtt_connection_cb(mqtt_client_t *client, void *arg, mqtt_connection_status_t status) {
+    MQTT_CLIENT_DATA_T* mqtt_client = (MQTT_CLIENT_DATA_T*)arg;
+    if (status == MQTT_CONNECT_ACCEPTED) {
+        mqtt_sub_unsub(client, "/led", 0, NULL, arg, 1);
+        printf("Connected to the /led topic successfully\n");
+    }
+}
+int main() {
+
+    stdio_init_all();
+
+    adc_init();
+    adc_set_temp_sensor_enabled(true);
+    adc_select_input(4);
+
+    mqtt = (MQTT_CLIENT_DATA_T*)calloc(1, sizeof(MQTT_CLIENT_DATA_T));
+    if (!mqtt) {
+        printf("Failed to inizialize MQTT client \n");
+        return 1;
+    }
+    // MQTT CLIENT INFO
+    mqtt->mqtt_client_info.client_id = MQTT_CLIENT_ID;
+    mqtt->mqtt_client_info.keep_alive = 60; // Keep alive in sec
+
+    if (cyw43_arch_init()) {
+        printf("Failed to inizialize CYW43\n");
+        return 1;
+    }
+
+    cyw43_arch_enable_sta_mode();
+    if (cyw43_arch_wifi_connect_timeout_ms(WIFI_SSID, WIFI_PASSWORD, CYW43_AUTH_WPA2_AES_PSK, 10000)) {
+        printf("Wi-Fi error\n");
+        return 1;
+    }
+    printf("\nConnected to Wifi\n");
+
+    ip_addr_t addr;
+    if (!ip4addr_aton(MQTT_BROKER_IP, &addr)) {
+        printf("MQTT ip Address not valid !\n");
+        return 1;
+    }
+
+    mqtt->mqtt_client_inst = mqtt_client_new();
+    if (!mqtt->mqtt_client_inst) {
+        printf("MQTT client instance creation error\n");
+        return 1;
+    }
+
+    if (mqtt_client_connect(mqtt->mqtt_client_inst, &addr, PORT, mqtt_connection_cb, mqtt, &mqtt->mqtt_client_info) != ERR_OK) {
+        printf("MQTT broker connection error\n");
+        return 1;
+    }
+    printf("Successfully connected to the MQTT broker\n");
+    mqtt_set_inpub_callback(mqtt->mqtt_client_inst, mqtt_incoming_publish_cb, mqtt_incoming_data_cb, mqtt);
+
+    while (1) {
+        publish_temperature(); // Public temperature every 5 sec
+        sleep_ms(5000);
+        tight_loop_contents();
```

Comment:
No need to do `tight_loop_contents()` if you're already doing `sleep_ms(5000)` on the previous line!